KAFKA-16689: Move LogValidatorTest to storage module

TaiJuWu opened the pull request:
*More detailed description of your change,
Rewrite LogValidatorTest.scala to java version

*Summary of testing strategy (including rationale)

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

TaiJuWu committed:
Wrtie logValidatorTest.scala to Java

TaiJuWu committed:
Add helper function iterToStream

TaiJuWu committed:
Remove LogValidatorTest.scala

TaiJuWu commented:
@chia7712 Please take look! Thanks.

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2100 @@
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class LogValidatorTest {
+    Time time = Time.SYSTEM;
```
please add `private final` modifiers

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2100 @@
+    Time time = Time.SYSTEM;
+    TopicPartition topicPartition = new TopicPartition("topic", 0);
+    Set<MetricName> metricsKeySet = KafkaYammerMetrics.defaultRegistry().allMetrics().keySet();
+    LogValidator.MetricsRecorder metricsRecorder = UnifiedLog.newValidatorMetricsRecorder(new BrokerTopicStats(false).allTopicsStats());
```
ditto

TaiJuWu replied:
Done.

TaiJuWu committed:
address comments

TaiJuWu committed:
Merge branch 'trunk' into KAFKA-16689

TaiJuWu committed:
Fix build error

TaiJuWu committed:
Merge branch 'trunk' into KAFKA-16689

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2103 @@
+            // Validation for v2 and above is strict for this case. For older formats, we fix invalid
+            // internal offsets by rewriting the batch.
+            if (version.value >= RecordBatch.MAGIC_VALUE_V2) {
+                assertThrows(InvalidRecordException.class, () -> {
```
please remove `{}`

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2103 @@
+
+        records.forEach(record -> {
+            try {
+                builder.appendUncheckedWithOffset(0, record);
```
Could you please use `assertDoesNotThrow` and then remove the "FIXME"?

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2103 @@
+        verifyRecordValidationStats(validatedResults.recordValidationStats, 0, records, true);
+    }
+
+    private MemoryRecords createRecords(byte magicValue,
```
please remove unused method

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2103 @@
+                compression
+        );
+
+        assertThrows(RecordValidationException.class, () -> {
```
please remove `{}`

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2103 @@
+                Compression.NONE
+        );
+
+        assertThrows(RecordValidationException.class, () -> {
```
please remove `{}`

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2103 @@
+                compression
+        );
+
+        assertThrows(RecordValidationException.class, () -> {
```
please remove `{}`

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2103 @@
+        Compression compression = Compression.gzip().build();
+        MemoryRecords records = createRecords(RecordBatch.MAGIC_VALUE_V2, now - 1001L, compression);
+
+        RecordValidationException e = assertThrows(RecordValidationException.class, () -> {
```
please remove `{}`

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2103 @@
+        }
+        MemoryRecords invalidOffsetTimestampRecords = builder.build();
+
+        RecordValidationException e = assertThrows(RecordValidationException.class, () -> {
```
please remove `{}`

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2103 @@
+        long fiveMinutesBeforeThreshold = now - timestampBeforeMaxConfig - (5 * 60 * 1000L);
+        Compression compression = Compression.gzip().build();
+        MemoryRecords records = createRecords(RecordBatch.MAGIC_VALUE_V2, fiveMinutesBeforeThreshold, compression);
+        RecordValidationException e = assertThrows(RecordValidationException.class, () -> {
```
please remove `{}`

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2103 @@
+        long fiveMinutesAfterThreshold = now + timestampAfterMaxConfig + (5 * 60 * 1000L);
+        Compression compression = Compression.gzip().build();
+        MemoryRecords records = createRecords(RecordBatch.MAGIC_VALUE_V2, fiveMinutesAfterThreshold, compression);
+        RecordValidationException e = assertThrows(RecordValidationException.class, () -> {
```
please remove `{}`

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2103 @@
+            );
+        });
+
+        assertTrue(e.invalidException() instanceof InvalidTimestampException);
```
please use `assertInstanceOf`

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2103 @@
+            );
+        });
+
+        assertTrue(e.invalidException() instanceof InvalidTimestampException);
```
please use `assertInstanceOf`

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2103 @@
+        });
+        // If there is a mix of both regular InvalidRecordException and InvalidTimestampException,
+        // InvalidTimestampException takes precedence
+        assertTrue(e.invalidException() instanceof InvalidTimestampException);
```
please use `assertInstanceOf`

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2103 @@
+            );
+        });
+
+        assertTrue(e.invalidException() instanceof InvalidTimestampException);
```
please use `assertInstanceOf`

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2103 @@
+        long offset = baseOffset;
+        Iterator<Record> iterator = records.records().iterator();
+
+        while (iterator.hasNext()) {
```
```java
      for (Record record : records.records()) {
        Assertions.assertEquals(offset, record.offset(), "Unexpected offset in message set iterator");
        offset += 1;
      }
```

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2103 @@
+    @Test
+    public void testDifferentLevelDoesNotCauseRecompression() {
+        List<byte[]> records = Arrays.asList(
+                Collections.nCopies(256, "some").stream().collect(Collectors.joining("")).getBytes(),
```
`String.join("", Collections.nCopies(256, "some")).getBytes(),`

chia7712 reviewed (commented):
@TaiJuWu thanks for updated PR

TaiJuWu committed:
address comments

TaiJuWu commented:
Hi @chia7712 , thanks for your review.
I have updated this PR, PTAL.

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2091 @@
+
+    @Test
+    public void testRecordWithPastTimestampIsRejected() {
+        long timestampBeforeMaxConfig = 24 * 60 * 60 * 1000L; // 24 hrs
```
how about using `Duration.ofHours(24).toMillis()`?

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2091 @@
+    @Test
+    public void testRecordWithPastTimestampIsRejected() {
+        long timestampBeforeMaxConfig = 24 * 60 * 60 * 1000L; // 24 hrs
+        long timestampAfterMaxConfig = 1 * 60 * 60 * 1000L; // 1 hr
```
ditto

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2091 @@
+        long timestampBeforeMaxConfig = 24 * 60 * 60 * 1000L; // 24 hrs
+        long timestampAfterMaxConfig = 1 * 60 * 60 * 1000L; // 1 hr
+        long now = System.currentTimeMillis();
+        long fiveMinutesBeforeThreshold = now - timestampBeforeMaxConfig - (5 * 60 * 1000L);
```
ditto

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2091 @@
+
+    @Test
+    public void testRecordWithFutureTimestampIsRejected() {
+        long timestampBeforeMaxConfig = 24 * 60 * 60 * 1000L; // 24 hrs
```
ditto

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2091 @@
+    @Test
+    public void testRecordWithFutureTimestampIsRejected() {
+        long timestampBeforeMaxConfig = 24 * 60 * 60 * 1000L; // 24 hrs
+        long timestampAfterMaxConfig = 1 * 60 * 60 * 1000L; // 1 hr
```
ditto

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2091 @@
+        long timestampBeforeMaxConfig = 24 * 60 * 60 * 1000L; // 24 hrs
+        long timestampAfterMaxConfig = 1 * 60 * 60 * 1000L; // 1 hr
+        long now = System.currentTimeMillis();
+        long fiveMinutesAfterThreshold = now + timestampAfterMaxConfig + (5 * 60 * 1000L);
```
ditto

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2091 @@
+     */
+    void validateLogAppendTime(long expectedLogAppendTime, long expectedBaseTimestamp, RecordBatch batch) {
+        assertTrue(batch.isValid());
+        assertTrue(batch.timestampType() == TimestampType.LOG_APPEND_TIME);
```
could you please use `assertEqual`?

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2091 @@
+        Assertions.assertTrue(iteratorSize(records.records().iterator()) != 0, "Message set should not be empty");
+
+        long offset = baseOffset;
+        Iterator<Record> iterator = records.records().iterator();
```
unused variable

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2091 @@
+            assertTrue(batch.isValid());
+            assertEquals(TimestampType.CREATE_TIME, batch.timestampType());
+            maybeCheckBaseTimestamp(timestampSeq.get(0), batch);
+            assertEquals(batch.maxTimestamp(), iterToStream(batch.iterator())
```
`iterToStream` is used to find the max timestamp, so we can enhance it to return max timestamp directly

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2091 @@
+        }
+    }
+
+    private MemoryRecords createNonIncreasingOffsetRecords(byte magicValue,
```
this is used by only one case, so we can inline it, right?

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2091 @@
+                PrimitiveRef.ofLong(0L), metricsRecorder, RequestLocal.withThreadConfinedCaching().bufferSupplier()
+        ));
+        assertEquals(metricsKeySet.stream().filter(m -> m.getMBeanName().endsWith(BrokerTopicStats.InvalidOffsetOrSequenceRecordsPerSec())).count(), 1);
+        assertTrue(kafka.utils.TestUtils$.MODULE$.meterCount(BrokerTopicStats.InvalidOffsetOrSequenceRecordsPerSec()) > 0);
```
Could you please avoid using scala TestUtils?

chia7712 replied:
Maybe we can create a impl of `MetricsRecorder` to collect the function call?

TaiJuWu replied:
Done. PTAL. This part is the last commit, please help to review, thanks.

chia7712 reviewed (commented):
@TaiJuWu thanks for updated PR

TaiJuWu committed:
Address comments
Address below comments
1. replace time with Duration
2. Delete iterToStream
3. remove unused variable
4. Inline createNonIncreasingOffsetRecords

TaiJuWu committed:
Add meterCount to LogValidator and remove scala TestUtils dependency

chia7712 commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogValidator.java:
```diff
@@ -63,6 +63,8 @@ public interface MetricsRecorder {
 
         void recordNoKeyCompactedTopic();
+        // Visible for test
+        Long meterCount(String metricName);
```
we don't need to change the interface. you can have a specific impl which can return `meterCount`

TaiJuWu replied:
@chia7712 thanks for review.
I have updated this PR. PTAL.

TaiJuWu committed:
Revert "Add meterCount to LogValidator and remove scala TestUtils dependency"

This reverts commit 51bffc7ef2cf81ca3a3ee787b2a8af50fb1f1d0f.

TaiJuWu committed:
remove scala TestUtils dependency

TaiJuWu committed:
Remove UnifiedLog and metric dependency

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -81,7 +81,30 @@ public class LogValidatorTest {
-    private final LogValidator.MetricsRecorder metricsRecorder = UnifiedLog.newValidatorMetricsRecorder(new BrokerTopicStats(false).allTopicsStats());
+    private final LogValidator.MetricsRecorder metricsRecorder = new MetricsRecorder();
+
+    static class MetricsRecorder implements LogValidator.MetricsRecorder {
```
Please don't copy production code to testing ...

TaiJuWu replied:
Sorry for this, updated already and I am working on new tests (we discuss offline.)

TaiJuWu committed:
testRecordInvalidSequence fail

The fail state (hasProducerId and sequence < 1) is already
banned when create memoryBuilder.

ref: https://github.com/apache/kafka/blob/638844f833b165d6f9ca52c173858d26b7254fac/clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java#L397

TaiJuWu committed:
add InvalidChecksum test

TaiJuWu committed:
Revert "testRecordInvalidSequence fail"

This reverts commit 19dc90285475dea4c440f085e65474d75ce2e5b2.

TaiJuWu committed:
add testNoKeyCompactedTopic test

TaiJuWu committed:
Merge branch 'trunk' into KAFKA-16689

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2177 @@
+            recordInvalidOffsetCount += 1;
+        }
+
+        public void recordInvalidSequence() {
```
Is it possible to add UT for this method?

TaiJuWu replied:
It was added, thanks.

TaiJuWu committed:
add testInvalidSequence

chia7712 commented on storage/src/main/java/org/apache/kafka/storage/internals/log/LogValidator.java:
```diff
@@ -574,7 +574,7 @@ private static Optional<ApiRecordError> validateRecord(RecordBatch batch,
             try {
                 record.ensureValid();
-            } catch (InvalidRecordException e) {
+            } catch (InvalidRecordException | CorruptRecordException e) {
```
why we need this check?

TaiJuWu replied:
[here](https://github.com/apache/kafka/blob/80f31224aad543dbfc892bce1ad73b6bb693855a/clients/src/main/java/org/apache/kafka/common/record/LegacyRecord.java#L129) is the exception should be thrown so we need to add it.

chia7712 replied:
My point was why we need to change the production code in this test migration?

TaiJuWu replied:
I can't enter `recordInvalidChecksums` if I don't have this modification.

chia7712 replied:
ok, it seems that is another issue.

#7150 changes the thrown exception from `InvalidRecordException` to `CorruptRecordException`, and so `LogValidator` can't catch it to update `recordInvalidChecksums`. Let me file a Jira for it

chia7712 replied:
@TaiJuWu I have assigned https://issues.apache.org/jira/browse/KAFKA-17104 to you

TaiJuWu replied:
Got it, thanks.

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2228 @@
+    }
+
+    @Test
+    public void testInvalidSequence() {
```
nice one. could you please use `EnumSource` to rewrite it? for example:
```java
    @ParameterizedTest
    @EnumSource(CompressionType.class)
    public void testInvalidSequenceV0(CompressionType type) {
        checkInvalidSequence(RecordBatch.MAGIC_VALUE_V0, type);
    }

    @ParameterizedTest
    @EnumSource(CompressionType.class)
    public void testInvalidSequenceV1(CompressionType type) {
        checkInvalidSequence(RecordBatch.MAGIC_VALUE_V2, type);
    }

    @ParameterizedTest
    @EnumSource(CompressionType.class)
    public void testInvalidSequenceV2(CompressionType type) {
        checkInvalidSequence(RecordBatch.MAGIC_VALUE_V2, type);
    }
```

TaiJuWu replied:
Hi @chia7712 , Thanks for review and nice suggestions.
I rewrite this part, please take a look.

chia7712 reviewed (commented):
@TaiJuWu thanks for updates

TaiJuWu committed:
rewrite testInvalidSequence by EnumSource

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2250 @@
+    }
+
+    @Test
+    public void testOnlyOneBatch() {
```
Could you please use `CsvSource` to rewrite them?

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2250 @@
+    }
+
+    @Test
+    public void testAllowMultiBatch() {
```
ditto

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2250 @@
+    }
+
+    @Test
+    public void testMisMatchMagic() {
```
ditto

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2250 @@
+    }
+
+    @Test
+    public void testLogAppendTimeNonCompressedV0() {
```
ditto

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2250 @@
+    }
+
+    @Test
+    void testNonCompressedV2() {
```
ditto

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2250 @@
+    }
+
+    @Test
+    public void testInvalidChecksum() {
```
ditto

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2250 @@
+        checkCompressed(RecordBatch.MAGIC_VALUE_V2);
+    }
+    @Test
+    void testInvalidOffsetRangeAndRecordCount() {
```
Could you please separate the different cases? for example:
```java
    @Test
    void testRecordBatchWithCountOverrides() {
        // The batch to be written contains 3 records, so the correct lastOffsetDelta is 2
        validateRecordBatchWithCountOverrides(2, 3);
    }

    @ParameterizedTest
    @CsvSource({"0,3", "15,3", "-3,3"})
    void testInconsistentCountAndOffset(int lastOffsetDelta, int count) {
        // Count and offset range are inconsistent or invalid
        assertInvalidBatchCountOverrides(lastOffsetDelta, count);
    }

    @ParameterizedTest
    @CsvSource({"5,6", "1,2"})
    void testUnmatchedNumberOfRecords(int lastOffsetDelta, int count) {
        // Count and offset range are consistent, but do not match the actual number of records
        assertInvalidBatchCountOverrides(lastOffsetDelta, count);
    }
```

TaiJuWu replied:
Done.

chia7712 commented on storage/src/test/java/org/apache/kafka/storage/internals/log/LogValidatorTest.java:
```diff
@@ -0,0 +1,2250 @@
+    }
+
+
+    private void checkInvalidSequence(byte magic, Compression compression, CompressionType type) {
```
Could you use `MethodSource`? for example:
```java
    private static Stream<Arguments> testInvalidSequence() {
        return Stream.of(RecordBatch.MAGIC_VALUE_V0, RecordBatch.MAGIC_VALUE_V1, RecordBatch.MAGIC_VALUE_V2)
                .flatMap(code -> Arrays.stream(CompressionType.values()).flatMap(source ->
                        Arrays.stream(CompressionType.values()).map(target -> Arguments.of(code, source, target))));
    }

    @ParameterizedTest
    @MethodSource("testInvalidSequence")
    void testInvalidSequence(byte magic, CompressionType source, CompressionType target) {
```

TaiJuWu replied:
Done.

chia7712 reviewed (commented):
@TaiJuWu thanks for your updates. There are many legacy style, and please update them by junit 5 style

TaiJuWu committed:
use junit5 parameterTest

TaiJuWu committed:
address comments

TaiJuWu committed:
checkNoKeyCompactedTopic use parameterTest

TaiJuWu committed:
increase test coverage

TaiJuWu committed:
use String in testInvalidSequenceArguments

TaiJuWu commented:
> @TaiJuWu thanks for your updates. There are many legacy style, and please update them by junit 5 style

@chia7712 thanks for your review and comments, I try updating to juint5 already, hope I don't miss anything.

chia7712 commented:
@TaiJuWu please fix the conflicts

TaiJuWu committed:
Merge branch 'trunk' into KAFKA-16689

chia7712 commented:
@TaiJuWu please rebase code to fix build error

TaiJuWu committed:
Merge branch 'trunk' into KAFKA-16689

TaiJuWu commented:
> @TaiJuWu please rebase code to fix build error

Done.

TaiJuWu committed:
fix build error

TaiJuWu committed:
remove scala LogValidatorTest

chia7712 commented:
@TaiJuWu Could you please merge trunk to fix those known failed tests?

TaiJuWu committed:
Merge branch 'trunk' into KAFKA-16689

chia7712 reviewed (approved):
LGTM
